brainiak/reconstruct/iem.py

```diff
@@ -0,0 +1,290 @@
+        pred_indx = np.argmax(pred_response, axis=1)
+        pred_dir = self.C_D_[pred_indx]
+        return pred_dir
+
+    def score(self, X, y):
+        """Calculate error measure of prediction.
+
+        Parameters
+        ----------
+            X: numpy matrix of voxel activation from new data
+                [observations,voxels]
+            y: numpy array of responses. [observations]
+
+        Returns
+        -------
+            rss: residual sum of squares of predicted
+                features compared to to actual features.
+        """
+        pred_dir = self.predict(X)
+        u = ((y - pred_dir)**2).sum()
+        v = ((y - np.mean(y))**2).sum()
+        rss = (1 - u/v)
+        return rss
+
+    def get_params(self, deep=True):
+        """Returns model parameters.
+
+        Parameters
+        ----------
+        deep: boolean. default true.
+            if true, returns params of sub-objects
+
+        Returns
+        -------
+        params: parameter of this object
+        """
+        return{"n_channels": self.n_channels,
+               "range_start": self.range_start,
+               "range_stop": self.range_stop}
+
+    def set_params(self, **parameters):
+        """Sets model parameters after initialization.
+
+        Parameters
+        ----------
+            params: structure with parameters and change values
+
+        Returns
+        -------
+            self.
+        """
+        for parameter, value in parameters.items():
+            setattr(self, parameter, value)
+        return self
+
+    def _define_channels(self):
+        """Define basis functions (aka channels).
+
+        Parameters
+        ----------
+            self.
+
+        Returns
+        -------
+            channels: numpy matrix of basis functions.
+                    [n_channels, function resolution].
+            channel_domain: numpy array of domain values.
+        """
+        channel_exp = 6
+        channel_density = 180
+        shifts = np.linspace(0,
+                             math.pi - math.pi/self.n_channels,
+                             self.n_channels)
+
+        channel_domain = np.linspace(self.range_start,
+                                     self.range_stop,
+                                     channel_density)
+
+        channels = np.zeros((self.n_channels, channel_density))
+        for i in range(self.n_channels):
+            channels[i, :] = np.cos(np.linspace(0, math.pi, channel_density)
+                                    - shifts[i]) ** channel_exp
+        # Check that channels provide sufficient coverage
+        ch_sum_range = np.max(np.sum(channels, 0)) - min(np.sum(channels, 0))
+        if ch_sum_range > np.deg2rad(self.range_stop - self.range_start)*0.1:
+            # if range of channel sum > 10% channel domain size
+            raise ValueError("Insufficient channel coverage.")
+        return channels, channel_domain
+
+    def _predict_channel_responses(self, X):
```

Comment:
Is this function necessary? It seems that the only role of `_predict_directions()` is to call this function, so I was wondering if it is more efficient to combine `_predict_channel_responses()` with `_predict_directions()`.

Reply:
This is a good suggestion, however, I wanted to keep the possibility, however remote, that someone might want to see what the channel responses are independently of just the prediction based on them. I noticed one of my lab mates doing so (although I think it was just a sanity check). If it's alright, I left them separate.

Reply:
I see! Thanks! That makes sense!